### ISSUE TITLE
rnn might be 7% slower than char-rnn?

### ISSUE THREAD

**Issue by hughperkins** [Element-Research/rnn#95]
Currently, per my only-just-completed-therefore-maybe-flawed prototyping, seems like element-research rnn is about 7% slower than Karpathy's char-rnn?

Hardware used:
- NVIDIA Geforce 940M

Iteration time, using Karpathy char-rnn, using cuda, 0.13s
```
27/21150 (epoch 0.064), train_loss = 3.29486426, grad/param norm = 4.2767e-01, time/batch = 0.1283s
28/21150 (epoch 0.066), train_loss = 3.28503671, grad/param norm = 4.0787e-01, time/batch = 0.1284s
```

(Actually, it used to be 0.11s for me, so something has changed in the last few months.  Logged issue for one commit that increases time by 0.01s on Karpathy's char-rnn for me https://github.com/karpathy/char-rnn/issues/145 )

Iteration time, using code in appendix below, element-research rnn, cuda, 0.14s
```
backward run=0.10964679718018
forward run=0.02655291557312
forward setup=0.0034151077270508
backward setup=0.00043725967407227
epoch=1	it=5/22307	seqLoss=176.9912955761	time=0.14084410667419
backward run=0.10971999168396
forward run=0.026543378829956
forward setup=0.003347635269165
backward setup=0.00041699409484863
epoch=1	it=6/22307	seqLoss=175.80709600449	time=0.14078497886658
```

Appendix, code to run char-rnn, using elment-research rnn:
```lua
-- copyright Hugh Perkins 2015
-- license MPLv2
-- this work contains many parts derived technically and conceptually from Andrej Karpathy's https://github.com/karpathy/char-rnn
-- which is under MIT license

require 'os'
require 'paths'
require 'torch'
require 'sys'
require 'nn'
require 'rnn'

function timer_init()
  local state = {}
  return state
--  state.time = sys.clock()
end

function timer_update(state, name)
  local now = sys.clock()
  if state.time ~= nil then
    local thisInterval = now - state.time
    state[state.name] = state[state.name] or 0
    state[state.name] = state[state.name] + thisInterval
  end
  state.time = now
  state.name = name
  return state
end

function timer_dump(state)
  for k,v in pairs(state) do
    if k ~= 'time' and k ~= 'name' then
      print(k .. '=' .. v)
    end
  end
end

local backend = 'cuda'
--backend = 'cl'
--backend = 'cpu'

if backend == 'cuda' then
  require 'cutorch'
  require 'cunn'
elseif backend == 'cl' then
  require 'cltorch'
  require 'clnn'
end

local dataset = 'tinyshakespeare'
local in_file = 'input.txt'
local in_t7 = 'input.t7'
local vocab_t7 = 'vocab.t7'
local weights_t7 = 'weights_$DATASET_$EPOCH_$IT.t7'

local dataDir = 'data/' .. dataset

local dropoutProb = 0.0
local hiddenSizes = {128, 128}
local learningRate = 0.04
local seqLength = 50
local batchSize = 50
local dumpIntervalIts = 50

local outDir = 'out'

-- for sampling, not used currently
function decode(ivocab, encoded)
  local str = ''
  for i=1,#encoded do
    local thisencoded = encoded[i]
    local thiscode = ivocab[thisencoded]
    str = str .. string.char(thiscode)
  end
  return str
end

-- not used currently
function encode_input(state, input_string, grow_vocab)
  local ivocab = state.ivocab
  local vocab = state.vocab
  local len = input_string:len()
  local input_coded = torch.ByteTensor(len)
  for i=1,len do
    local char = input_string:byte(i)
    if vocab[char] == nil then
      if grow_vocab then
        ivocab[#ivocab + 1] = char
        vocab[char] = #ivocab
      else
        error('Unrecognized char ' .. input_string[i])
      end
    end
    local v = vocab[char]
    input_coded[i] = v
  end
  return input_coded
end

function getFileSize(filePath)
  local sizestr = sys.execute('stat --format=%s ' .. filePath)
  local size = tonumber(sizestr)
  return size
end

-- shamelessly adapted from Karpathy's char-rnn :-)
function text_to_t7(in_textfile, out_tensorfile, out_vocabfile)
  local cache_len = 10000
  local rawdata
  local tot_len = getFileSize(in_textfile)
  print('tot_len', tot_len)

  local input_coded = torch.ByteTensor(tot_len)
  local vocab = {}
  local ivocab = {}
  local state = {}
  state.vocab = vocab
  state.ivocab = ivocab

  local f = io.open(in_textfile, "r")
  local input_string = f:read(cache_len)
  local pos = 1
  repeat
    local len = input_string:len()
    print('len', len)
    for i=1,len do
      local char = input_string:byte(i)
      if vocab[char] == nil then
        ivocab[#ivocab + 1] = char
        vocab[char] = #ivocab
      end
      local v = vocab[char]
      input_coded[pos] = v
      pos = pos + 1
    end
    input_string = f:read(cache_len)
  until not input_string
  f:close()

  print('#ivocab', #ivocab)
  local vocabs = {}
  vocabs.vocab = vocab
  vocabs.ivocab = ivocab
  torch.save(out_vocabfile, vocabs)
  torch.save(out_tensorfile, input_coded)
end

if not paths.dirp(outDir) then
  paths.mkdir(outDir)
end

if not paths.filep(dataDir .. '/' .. in_t7) or not paths.filep(dataDir .. '/' .. vocab_t7) then
  text_to_t7(dataDir .. '/' .. in_file, dataDir .. '/' .. in_t7, dataDir .. '/' .. vocab_t7)
--  local f = io.open(dataDir + '/' + in_file, 'r')
  
--  f:close()
end

function getLabel(outputs)
  _, i = outputs:max(1)
  return i[1]
end

function padleft(pad, val)
  local str = '' .. val
  while str:len() < pad do
    str = ' ' .. str
  end
  return str
end

-- resulting input will be one less long than input_string
-- since we need one additional character for last target
function string_to_input_targets(state, input_string)
  local encoded_string = encode_input(state, input_string, true)
  local len = encoded_string:size(1)
  local inputs = {}
  local targets = {}
  for i=1,len-1 do
    local this_tensor = torch.Tensor(#ivocab):zero()  -- one-hot
    local v = encoded_string[i]
    this_tensor[encoded_string[i]] = 1
    table.insert(inputs, this_tensor)
    table.insert(targets, encoded_string[i+1])
  end
  local results = {}
  results.inputs = inputs
  results.targets = targets
  return results
end

function one_hot(size, value)
  local tensor = torch.Tensor(size):zero()
  tensor[value] = 1
  return tensor
end

-- resulting input will be same length as input_string
-- (cf string_to_input_targets)
function string_to_input(state, input_string)
  local encoded_string = encode_input(state, input_string, false)
  local len = encoded_string:size(1)
  local inputs = {}
  for i=1,len do
    local this_tensor = torch.Tensor(#ivocab):zero()  -- one-hot
    local v = encoded_string[i]
    this_tensor[encoded_string[i]] = 1
    table.insert(inputs, this_tensor)
  end
  return inputs
end

function sample(state, net, prefix, length)
  local res = ''
  local ivocab = state.ivocab
  local inputs = string_to_input(state, prefix)
  net:evaluate()
  net:forget()
  local lastEncoded = nil
  local lastChar = nil
  for i=1,#inputs do
    local input = inputs[i]
    local output = net:forward(input)
    lastEncoded = getLabel(output)
    lastChar = string.char(ivocab[lastEncoded])
    res = res .. lastChar
  end

  for i=1,length do
    local input = one_hot(#ivocab, lastEncoded)
    local output = net:forward(input)
    lastEncoded = getLabel(output)
    lastChar = string.char(ivocab[lastEncoded])
    res = res .. lastChar
  end
  print('sample', prefix, res)
  return res
end

local vocabs = torch.load(dataDir .. '/' .. vocab_t7)
local input = torch.load(dataDir .. '/' .. in_t7)
print('vocabs', vocabs)
for k,v in pairs(vocabs) do
  print('k', k)
end
print('loaded input')
local ivocab = vocabs.ivocab
local vocab = vocabs.vocab

local net = nn.Sequential()
local inputSize = #ivocab
for i, hiddenSize in ipairs(hiddenSizes) do
  net:add(nn.FastLSTM(inputSize, hiddenSize))
--  lm:add(nn.Sequencer(nn.Dropout(opt.dropoutProb)))
  inputSize = hiddenSize
end

net:add(nn.Linear(inputSize, #ivocab))
net:add(nn.LogSoftMax())

local crit = nn.ClassNLLCriterion()

print('#ivocab', #ivocab)

local batchInput = torch.Tensor(batchSize, #ivocab)
if backend == 'cuda' then
  input = input:cuda()
  batchInput = batchInput:cuda()
  net:cuda()
  crit:cuda()
elseif backend == 'cl' then
  input = input:cl()
  batchInput = batchInput:cl()
  net:cl()
  crit:cl()
else
end

local input_len = input:size(1)
local batchSpacing = math.ceil(input_len / batchSize)
local inputDouble = input:clone():resize(2, input_len)
inputDouble[1]:copy(input)
inputDouble[2]:copy(input)
inputDouble = inputDouble:reshape(input_len * 2)
local inputStriped = inputDouble:unfold(1, input_len, batchSpacing):t()

local it = 1
local epoch = 1
--local it = 1
local itsPerEpoch = math.floor(input:size(1) / seqLength)
local params = net:getParameters()
net:training()
net:backwardOnline()
while true do
  sys.tic()
  local seqLoss = 0
  net:forget()
  net:zeroGradParameters()
  batchInputs = {}
  batchOutputs = {}

  local timer = timer_init()
  for s=1,seqLength do
    local batchOffset = ((epoch - 1) + (it - 1) * seqLength + (s - 1) + 1 - 1) % input_len + 1

    timer_update(timer, 'forward setup')
    local bc2 = inputStriped[batchOffset]
    batchInput:zero()
    batchInput:scatter(2, bc2:reshape(batchSize, 1), 1)

    timer_update(timer, 'forward run')
    local batchOutput = net:forward(batchInput)
    batchInputs[s] = batchInput
    batchOutputs[s] = batchOutput
  end

  for s=seqLength,1,-1 do
    timer_update(timer, 'backward setup')
    local batchOffset = ((epoch - 1) + (it - 1) * seqLength + (s - 1) + 1 - 1) % input_len + 1
    local bt2 = inputStriped[batchOffset + 1]

    timer_update(timer, 'backward run')
    local batchLoss = crit:forward(batchOutputs[s], bt2)
    seqLoss = seqLoss + batchLoss
    local batchGradOutput = crit:backward(batchOutputs[s], bt2)
    net:backward(batchInputs[s], batchGradOutput)
  end
  timer_update(timer, 'end')
  timer_dump(timer)

  net:updateParameters(learningRate)
  print('epoch=' .. epoch, 'it=' .. it .. '/' .. itsPerEpoch, 'seqLoss=' .. seqLoss, 'time=' .. sys.toc())
  if it ~= 1 and ((it - 1) % dumpIntervalIts == 0) then
    local filename = weights_t7:gsub('$DATASET', dataset):gsub('$EPOCH', epoch):gsub('$IT', it)
    print('filename', filename)
    torch.save(outDir .. '/' .. filename, params) 
  end
  it = it + 1
  if it > itsPerEpoch then
    epoch = epoch + 1
    it = 1
  end
end
```

(edited to add missing `batchInput:zero()`.  Timings unchanged)

**Comment by nicholas-leonard** [comment]
@hughperkins So I just did a little benchmark with [updated code](https://github.com/hughperkins/char-rnn-er/pull/4) using rnn instead of nngraph. With those optimizations I get `0.003563 sec/step` (using cuda). This metric is directly comparable to char-rnn's `time/batch` (they define a batch as 50 examples forward/backward/updated for one time-step). So that would make our implementation 36 times faster then char-rnn's... 

I still need to confirm that they don't use peep-hole connections with their LSTM, because FastLSTM doesn't. Just in case, I ran the same script with nn.LSTM instead of nn.FastLSTM and I get `0.006654 sec/step`. Which is still 19 times faster. 

I think most of this speedup is due to better memory management on our part. The char-rnn code is unnecessarily re-allocating buffers for every batch (see their [utils/OneHot.lua code](https://github.com/karpathy/char-rnn/blob/master/util/OneHot.lua#L16-L18) or their [train.lua:prepro function](https://github.com/karpathy/char-rnn/blob/master/train.lua#L205-L219)).  

In any case, this benchmark is meaningless until we can use our implementation to get the same NLL convergence as them. Working on it.

**Comment by hughperkins** [comment]
> So I just did a little benchmark with updated code using rnn instead of nngraph. With those optimizations I get 0.003563 sec/step (using cuda).

This message appeared in my email, but not in the thread ;-)  I guess that somehow this was using a batchsize of 1, instead of 50?

**Comment by nicholas-leonard** [comment]
@hughperkins So I was able to get the difference down to `112 ms` (rnn) vs `110 ms` (char-rnn). So the slowdown is now about  `1.8 %` which I think is totally acceptable considering char-rnn being an incredibly small network (few parameters) and rnn being a much more flexible framework than char-rnn.
